pkg/cmd/secret/set/set.go

```diff
@@ -0,0 +1,216 @@
+	Visibility      string
+	RepositoryNames []string
+}
+
+func NewCmdSet(f *cmdutil.Factory, runF func(*SetOptions) error) *cobra.Command {
+	opts := &SetOptions{
+		IO:         f.IOStreams,
+		HttpClient: f.HttpClient,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "set <secret name>",
+		Short: "Create or update secrets",
+		Long:  "Locally encrypt a new or updated secret at either the repository or organization level and send it to GitHub for storage.",
+		Example: heredoc.Doc(`
+			$ cat SECRET.txt | gh secret set NEW_SECRET
+			$ gh secret set NEW_SECRET -b"some literal value"
+			$ gh secret set NEW_SECRET -b"@file.json"
+			$ gh secret set ORG_SECRET --org
+			$ gh secret set ORG_SECRET --org=anotherOrg --visibility=selected -r="repo1,repo2,repo3"
+			$ gh secret set ORG_SECRET --org=anotherOrg --visibility="all"
+`),
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return &cmdutil.FlagError{Err: errors.New("must pass single secret name")}
+			}
+			if !cmd.Flags().Changed("body") && opts.IO.IsStdinTTY() {
+				return &cmdutil.FlagError{Err: errors.New("no --body specified but nothing on STIDN")}
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// support `-R, --repo` override
+			opts.BaseRepo = f.BaseRepo
+
+			opts.SecretName = args[0]
+
+			err := validSecretName(opts.SecretName)
+			if err != nil {
+				return err
+			}
+
+			if cmd.Flags().Changed("visibility") {
+				if opts.OrgName == "" {
+					return &cmdutil.FlagError{Err: errors.New(
+						"--visibility not supported for repository secrets; did you mean to pass --org?")}
+				}
+
+				if opts.Visibility != shared.VisAll && opts.Visibility != shared.VisPrivate && opts.Visibility != shared.VisSelected {
+					return &cmdutil.FlagError{Err: errors.New(
+						"--visibility must be one of `all`, `private`, or `selected`")}
+				}
+			}
+
+			if cmd.Flags().Changed("repos") && opts.Visibility != shared.VisSelected {
+				return &cmdutil.FlagError{Err: errors.New(
+					"--repos only supported when --visibility='selected'")}
+			}
+
+			if opts.Visibility == shared.VisSelected && len(opts.RepositoryNames) == 0 {
+				return &cmdutil.FlagError{Err: errors.New(
+					"--repos flag required when --visibility='selected'")}
+			}
+
+			if runF != nil {
+				return runF(opts)
+			}
+
+			return setRun(opts)
+		},
+	}
+	cmd.Flags().StringVar(&opts.OrgName, "org", "", "List secrets for an organization")
+	cmd.Flags().Lookup("org").NoOptDefVal = "@owner"
+	cmd.Flags().StringVarP(&opts.Visibility, "visibility", "v", "private", "Set visibility for an organization secret: `all`, `private`, or `selected`")
+	cmd.Flags().StringSliceVarP(&opts.RepositoryNames, "repos", "r", []string{}, "List of repository names for `selected` visibility")
+	cmd.Flags().StringVarP(&opts.Body, "body", "b", "-", "Provide either a literal string or a file path; prepend file paths with an @. Reads from STDIN if not provided.")
```

Comment:
Is there a reason, or precedence, for using `@` to delineate a file path? I am wondering if it might make more sense to just introduce a flag for file input.

Reply:
It's based on `gh api`'s magic fields